lib/remap-functions/src/encode_json.rs

```diff
@@ -0,0 +1,154 @@
+use remap::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct EncodeJson;
+
+impl Function for EncodeJson {
+    fn identifier(&self) -> &'static str {
+        "encode_json"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |_| true,
+            required: true,
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+
+        Ok(Box::new(EncodeJsonFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct EncodeJsonFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for EncodeJsonFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let value = self.value.execute(state, object)?;
+        serde_json::to_string(&value)
+            .map(|v| v.into())
+            .map_err(|e| format!("unable encode to json: {}", e).into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .into_fallible(true) // Serialization error
+            .with_constraint(value::Kind::Bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use chrono::{DateTime, Utc};
+    use regex::Regex;
+    use value::Kind;
+
+    test_function![
+        encode_json => EncodeJson;
+
+        bytes {
+            args: func_args![value: r#"hello"#],
+            want: Ok(r#""hello""#),
+        }
+
+        integer {
+            args: func_args![value: 42],
+            want: Ok("42"),
+        }
+
+        float {
+            args: func_args![value: 42f64],
+            want: Ok("42.0"),
+        }
+
+        boolean {
+            args: func_args![value: false],
+            want: Ok("false"),
+        }
+
+        map {
+            args: func_args![value: map!["field": "value"]],
+            want: Ok(r#"{"field":"value"}"#),
+        }
+
+        array {
+            args: func_args![value: vec![1, 2, 3]],
+            want: Ok("[1,2,3]"),
+        }
+
+        timestamp {
+            args: func_args![
+                value: DateTime::parse_from_str("1983 Apr 13 12:09:14.274 +0000", "%Y %b %d %H:%M:%S%.3f %z")
+                    .unwrap()
+                    .with_timezone(&Utc)
+            ],
+            want: Ok(r#""1983-04-13 12:09:14.274 UTC""#),
```

Comment:
I would have expected an RFC3339 timestamp here; which format is this?

I could also see us wanting to allow users to configure the serialization format of timestamps, but we can add that later.

Reply:
This `Display` for `DateTime<UTC>`... if we want to change serialization format we need to change it in `remap-lang`:
https://github.com/timberio/vector/blob/507caed13f75973a435fe6b27417795c2eaee421/lib/remap-lang/src/value.rs#L45
(of course, we can use custom format here too, but a better would be to change in lib?)
`Display` is `Date Time TZ`:
`Date Time`: https://docs.rs/chrono/0.4.19/src/chrono/naive/datetime.rs.html#1447-1451
`Date Time TZ`: https://docs.rs/chrono/0.4.19/src/chrono/datetime.rs.html#726-733

Reply:
Ah I see, hmm. I think RFC3339 timestamps would be more expected here. I think we should change it in `remap-lang` to just do `serializer.serialize_str(&v)` as `DateTime` does implement `Serialize` (assuming we depend on the `serde` feature of `chrono`) which gives the expected RFC3339 timestamp.

Reply:
@jszwedko I created issue for this change: https://github.com/timberio/vector/issues/5677